### PR TITLE
[NCL-8169] Handle errors properly

### PR DESCRIPTION
**Pull request by patrikk0123**

### CODE REVIEW

**Review by matedo1 — CHANGES_REQUESTED** [review]

File: src/hooks/useServiceContainer.ts [hunk 1]
```diff
@@ -146,3 +141,17 @@ export const useServiceContainer = (service: Function, loadingStateDelayMs: numb
...
+
+const getErrorMessage = (error: Error | AxiosError): string => {
+  if (isAxiosError(error)) {
+    if (error.code === 'ERR_NETWORK') {
+      return 'Action was not successful due to the network error. Please, try again.';
+    } else if (error.response?.status === 401) {
+      return 'Action was not successful, please login first and try again. [HTTP 401 Unauthorized]';
+    }
+
+    return error.response?.data?.errorMessage ? error.response.data.errorMessage : error.toString();
```
Comment (matedo1) [inline src/hooks/useServiceContainer.ts:153]: Just a question to consider for the future - what if we need to adjust this message to provide some more specific information from the UI side.

Next, can we provide some general prefix like `Actions was no successful.` ? At this point we are able to say http request resulted in error which means action should be always failed.

File: src/common/types.ts [hunk 2]
```diff
@@ -1,3 +1,14 @@
 export type WithRequiredProperty<Type, Key extends keyof Type> = Type & {
   [Property in Key]-?: Type[Property];
 };
+
+type ERROR_CODE = 'NEW_ENTITY_ID_ERROR';
+
+export class CustomError extends Error {
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/common/types.ts]: Done, except for values like `NEW_ENTITY_ID_ERROR` being part of the `PncError` namespace. Instead, they are part of `PncErrorCode` namespace.

An array of those string values can be created, and from it, they can be added as properties to `PncError` class, but unfortunately, TypeScript does not recognize it. So all those value properties would need to be specified manually.

File: src/hooks/useServiceContainer.ts [hunk 3]
```diff
@@ -146,3 +141,17 @@ export const useServiceContainer = (service: Function, loadingStateDelayMs: numb
     run: useCallback(serviceContainerRunner, [service, loadingStateDelayMs]),
   };
 };
+
+const getErrorMessage = (error: Error | AxiosError): string => {
+  if (isAxiosError(error)) {
+    if (error.code === 'ERR_NETWORK') {
+      return 'Action was not successful due to the network error. Please, try again.';
+    } else if (error.response?.status === 401) {
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/hooks/useServiceContainer.ts]: done

File: src/hooks/useServiceContainer.ts [hunk 4]
```diff
@@ -146,3 +141,17 @@ export const useServiceContainer = (service: Function, loadingStateDelayMs: numb
     run: useCallback(serviceContainerRunner, [service, loadingStateDelayMs]),
   };
 };
+
+const getErrorMessage = (error: Error | AxiosError): string => {
+  if (isAxiosError(error)) {
+    if (error.code === 'ERR_NETWORK') {
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/hooks/useServiceContainer.ts]: done

File: src/hooks/useForm.ts [hunk 5]
```diff
@@ -250,11 +255,17 @@ export const useForm = () => {
...
       }
 
-      onSubmit(transformFormToValues(fields));
-
       setFields(fieldsCopy);
       setHasFormChanged(false);
       setIsSubmitDisabled(true);
+
+      onSubmit(transformFormToValues(fields)).catch((error: AxiosError | CustomError) => {
+        if (error.code === 'ERR_NETWORK') {
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/hooks/useForm.ts]: both done

File: src/hooks/useServiceContainer.ts [hunk 6]
```diff
@@ -146,3 +141,17 @@ export const useServiceContainer = (service: Function, loadingStateDelayMs: numb
...
+      return 'Action was not successful due to the network error. Please, try again.';
+    } else if (error.response?.status === 401) {
+      return 'Action was not successful, please login first and try again. [HTTP 401 Unauthorized]';
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/hooks/useServiceContainer.ts:152]: I am not sure, what do you mean. Where should the `errorMessage` from the screenshot of yours be defined?

File: src/hooks/useServiceContainer.ts (hunk 1 above)
Comment (patrikk0123) [reply src/hooks/useServiceContainer.ts:155]: I think `Actions was no successful` is not all that much needed and the top-level error message is obvious enough.
Additional parsing, I think, might be considered in the future if such need arises.

File: src/hooks/useForm.ts [hunk 7]
```diff
@@ -250,11 +255,17 @@ export const useForm = () => {
...
+          setIsSubmitDisabled(false);
+        } else if (error.code === 'NEW_ENTITY_ID_ERROR') {
+          uiLogger.error(error.message);
```
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/hooks/useForm.ts]: Unfortunately, `PncError` cannot be caught in the `pncClient` or its inceptor for it is thrown after the service execution.

But I would consider testing the ID always in the HTTP POST request interceptor. This way code repetition would be reduced. But this has one issue: POST is not _always_ used as CREATE action.

File: src/common/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/common/types.ts]: Done 

During the demo meeting we verified list of `ERROR_CODE`s is automatically provided in nested `isPncError` sections.

File: src/hooks/useForm.ts [hunk 8]
```diff
@@ -250,11 +255,21 @@ export const useForm = () => {
...
+      onSubmit(transformFormToValues(fields)).catch((error: AxiosError | PncError) => {
+        if (isAxiosError(error)) {
+          if (error.code === AxiosError.ERR_NETWORK) {
+            setIsSubmitDisabled(false);
+          }
+        } else if (isPncError(error)) {
+          if (error.code === 'NEW_ENTITY_ID_ERROR') {
+            uiLogger.error(error.message);
+          }
+        }
```
Comment (matedo1) [inline src/hooks/useForm.ts:271]: Missing final `else`, are we sure no other `Error` can happen?

File: src/hooks/useServiceContainer.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/hooks/useServiceContainer.ts]: We discussed it during the demo meeting.

File: src/hooks/useForm.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/hooks/useForm.ts]: For now we agreed to keep it as it is, as only manually defined http responses needs to be checked and there is probably no other better placed to do it.

File: src/hooks/useServiceContainer.ts (hunk 1 above)
Comment (matedo1) [reply src/hooks/useServiceContainer.ts]: Agreed

File: src/hooks/useServiceContainer.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/hooks/useServiceContainer.ts]: We discussed it during the demo meeting.

File: src/hooks/useForm.ts (hunk 8 above)
Comment (matedo1) [inline src/hooks/useForm.ts:271]: Missing final `else`, are we sure no other `Error` can happen?

**Review by matedo1 — APPROVED** [review]
Approved, just one question.

File: src/hooks/useForm.ts (hunk 8 above)
Comment (patrikk0123) [reply src/hooks/useForm.ts:271]: Quite much so, but I put there uiLogger logging just in case.

**Pull request merged**